src/ppom-storage.ts

```diff
@@ -0,0 +1,208 @@
+ */
+type FileInfo = {
+  name: string;
+  chainId: string;
+  version: string;
+  checksum: string;
+};
+
+/**
+ * PPOMFileMetadata Type
+ * This is type of metadata about files saved in storage,
+ * this information is saved in PPOMController state.
+ */
+export type PPOMFileMetadata = FileInfo[];
+
+/**
+ * StorageKey Type
+ * This defines type of key that is used for indexing file data saved in StorageBackend.
+ */
+export type StorageKey = {
+  name: string;
+  chainId: string;
+};
+
+/**
+ * StorageBackend Type
+ * This defines type for storage backend implementation.
+ * There will be different storage implementations depending on platform:
+ * 1. extension - indexDB
+ * 2. mobile app - <TBD>
+ */
+export type StorageBackend = {
+  read(key: StorageKey, checksum: string): Promise<ArrayBuffer>;
+  write(key: StorageKey, data: ArrayBuffer, checksum: string): Promise<void>;
+  delete(key: StorageKey): Promise<void>;
+  dir(): Promise<StorageKey[]>;
+};
+
+/**
+ * PPOMStorage class
+ * This class is responsible for managing the local storage
+ * It provides the following functionalities:
+ * 1. Sync the metadata with the version info from the cdn
+ * 2. Read a file from the local storage
+ * 3. Write a file to the local storage
+ *
+ * It also validates the checksum of the file when reading and writing in order to
+ * detect corrupted files or files that are not up to date
+ */
+export class PPOMStorage {
+  readonly #storageBackend: StorageBackend;
+
+  readonly #readMetadata: () => PPOMFileMetadata;
+
+  readonly #writeMetadata: (metadata: PPOMFileMetadata) => void;
+
+  /**
+   * Creates a PPOMStorage instance.
+   *
+   * @param options - The options passed to the function.
+   * @param options.storageBackend - The storage backend to use for the local storage.
+   * @param options.readMetadata - A function to read the metadata from the local storage.
+   * @param options.writeMetadata - A function to write the metadata to the local storage.
+   */
+  constructor({
+    storageBackend,
+    readMetadata,
+    writeMetadata,
+  }: {
+    storageBackend: StorageBackend;
+    readMetadata: () => PPOMFileMetadata;
+    writeMetadata: (metadata: PPOMFileMetadata) => void;
+  }) {
+    this.#storageBackend = storageBackend;
+    this.#readMetadata = readMetadata;
+    this.#writeMetadata = writeMetadata;
+  }
+
+  /**
+   * Sync the metadata with the version info from the cdn.
+   * 1. Remove the files that are not readable (e.g. corrupted or deleted).
+   * 2. Remove the files that are not in the cdn anymore.
+   * 3. Remove the files that are not up to date in the cdn.
+   * 4. Remove the files that are not in the local storage from the metadata.
+   * 5. Delete the files that are not in the metadata from the local storage.
+   *
+   * @param versionInfo - Version information of metadata files.
+   */
+  async syncMetadata(versionInfo: FileInfo[]): Promise<PPOMFileMetadata> {
```

Comment:
Should the type of `versionInfo` also be `PPOMFileMetadata`?

Reply:
Good point, I fixed it.